### DISCUSSION TITLE
How to use it on wasm?

### DISCUSSION THREAD

**Discussion by bajrangCoder** [Eugeny/russh#382] [Q&A] [answered]
After checking the repository and PRs, I see that the library supports WASM. However, when I tried using it to create a very basic SSH client, I ran into issues during the compilation to WASM. The problem seems to stem from `tokio`, as it relies on certain features that are not supported in a WASM environment.  

The error I encountered was something like this:  

```
error: Only features sync, macros, io-util, rt, time are supported on wasm.
```

I would really appreciate it if there were any examples or resources available related to using this library with WASM.  

For reference, I am using version 0.46 of `russh`.

**Comment by bajrangCoder** [comment]
@irvingoujAtDevolution You should also include a simple wasm example

**Comment by irvingoujAtDevolution** [comment]
You cannot use tokio, try [wasm_bindgen_futures](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen_futures/)

**Reply by bajrangCoder** [reply]
Yes, I know . I'm not using tokio in my project just russh, russh-keys , wasm_bindgen etc (I'm not using tokio)

**Comment by psychon** [comment]
Not really helpful, but: It just worked for me when I tried it. Sorry, I cannot share an example, but... where does your dependency on tokio come from? Are _you_ perhaps enabling some tokio features that you must not enable?

**Reply by bajrangCoder** [reply]
No, maybe some small reference or example will be appreciated

**Comment by irvingoujAtDevolution** [comment]
I think I know your problem, did you use connect? not connect_stream?

**Reply by bajrangCoder** [reply]
Yes, I'm using connect

**Reply by irvingoujAtDevolution** [reply, marked as answer]
https://github.com/irvingoujAtDevolution/russh-wasm-example

**Reply by irvingoujAtDevolution** [reply]
That's why, here is the example you looking for. Sorry I wanted to do it but forget it afterward. You need to use connect_stream and relay your websocket connection with a relay server. This repo supports wasm, but not wasi, so no tcp connection allowed.

**Reply by bajrangCoder** [reply]
Thanks 👍 

Btw it will be nice to have wasi support

**Comment by ankurtechno** [comment]
Hi, I am trying to build it, this sample is working correctly but i need to use russh-keys and if i include it i get error: This wasm target is unsupported by mio. If using Tokio, disable the net feature. Can you please help what is needed here

**Reply by psychon** [reply]
You have something in your dependency tree that enables the `net` feature on tokio (or are enabling it yourself). You have to get rid of that.

I had that problem was as well and had to split some test-only dependency out so that it is not used for the wasm target. Sadly, I do not remember the details nor how I debugged this. But basically: Do what the error message says and do not enable tokio's `net` feature. :-)

**Reply by ankurtechno** [reply]
Thank you, I found out it was russh-keys version, I used 0.49.2 and it worked correctly.

**Comment by ankurtechno** [comment]
I have one more question, when we use this as web assembly, how is internal threading works. Is there any set of rules we need to follow.

**Comment by ankurtechno** [comment]
I am trying to use sftp with wasm, call to   let sftp = SftpSession::new(sftp_channel.into_stream()).await.unwrap(); is causing panic, here is the stack   at russh_client_core.wasm.tokio::task::spawn::spawn::h56104c0206615506 (russh_client_core_bg.wasm:0x1a28dd)
    at russh_client_core.wasm.russh_sftp::client::run::h67bf193394f44162 (russh_client_core_bg.wasm:0x18df97).

What needs to be done here?
